src/Grpc.Net.Client/Internal/Retry/RetryCall.cs

```diff
@@ -0,0 +1,320 @@
+
+                        _nextRetryDelayMilliseconds = CalculateNextRetryDelay();
+
+                        // Check if dispose was called on call.
+                        CancellationTokenSource.Token.ThrowIfCancellationRequested();
+
+                        // Clean up the failed call.
+                        currentCall.Dispose();
+                    }
+                    else
+                    {
+                        // Handle the situation where the call failed with a non-deadline status, but retry
+                        // didn't happen because of deadline exceeded.
+                        IGrpcCall<TRequest, TResponse> resolvedCall = (IsDeadlineExceeded() && !(currentCall.CallTask.IsCompletedSuccessfully() && currentCall.CallTask.Result.StatusCode == StatusCode.DeadlineExceeded))
+                            ? CreateStatusCall(GrpcProtocolConstants.DeadlineExceededStatus)
+                            : currentCall;
+
+                        // Can't retry.
+                        // Signal public API exceptions that they should finish throwing and then exit the retry loop.
+                        CommitCall(resolvedCall, result.GetValueOrDefault());
+                        return;
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                HandleUnexpectedError(ex);
+            }
+            finally
+            {
+                Log.StoppingRetryWorker(Logger);
+            }
+        }
+
+        protected override void OnCommitCall(IGrpcCall<TRequest, TResponse> call)
+        {
+            _activeCall = null;
+        }
+
+        protected override void Dispose(bool disposing)
+        {
+            lock (Lock)
+            {
+                base.Dispose(disposing);
+
+                _activeCall?.Dispose();
+            }
+        }
+
+        protected override void StartCore(Action<GrpcCall<TRequest, TResponse>> startCallFunc)
+        {
+            _ = StartRetry(startCallFunc);
+        }
+
+        public override Task ClientStreamCompleteAsync()
+        {
+            ClientStreamComplete = true;
+
+            return DoClientStreamActionAsync(async call =>
+            {
+                await call.ClientStreamWriter!.CompleteAsync().ConfigureAwait(false);
+            });
+        }
+
+        public override Task ClientStreamWriteAsync(TRequest message)
+        {
+            return DoClientStreamActionAsync(async call =>
+            {
+                CompatibilityExtensions.Assert(call.ClientStreamWriter != null);
+
+                if (ClientStreamWriteOptions != null)
+                {
+                    call.ClientStreamWriter.WriteOptions = ClientStreamWriteOptions;
+                }
+
+                await call.WriteClientStreamAsync(WriteNewMessage, message).ConfigureAwait(false);
+                BufferedCurrentMessage = false;
```

Comment:
Hmm this value is read while holding the lock, should the lock be held when writing it?